### PR TITLE
docs: add fe coding guidelines

### PR DESCRIPTION
**Pull request by joshuaellis**
### What does it do?

* adds FE coding guidelines from the V5 RFC

### CODE REVIEW

File: docs/docs/guides/04-fe-coding-guidelines.mdx [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+import { useState } from 'react';
+
+const App = () => {
+  const [someState, setSomeState] = useState('some state');
+  return (
+    <>
+      <Header
+        logo={<Logo someState={someState} />}
+        settings={<Settings onStateChange={setSomeState} />}
+      />
```
Comment (markkaylor) [inline docs/docs/guides/04-fe-coding-guidelines.mdx:334]: Don't we want to encourage this pattern? Or maybe you are trying to illustrate multiple options?

```suggestion
      <Header>
      	<Logo someState={someState} />
      	<Settings onStateChange={setSomeState} />
      </Header>
```

File: docs/docs/guides/04-fe-coding-guidelines.mdx [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+past that point, they're organic and re-organise as we continue to grow the documentation. The `API`
+section at time of writing is light and only discusses the `Strapi` class. Moving forward, we'd like
+to include all the `admin` API endpoints and their return values.
+
+#### When should you add content?
+
+Content should be added typically when you add any new feature – especially enterprise e.g.
+`Review Workflows` and also when you add new code other engineers may find another use for e.g. a
+new hook. There's no guidelines on what to add vs what not to add, as defined above we add a variety
+of material to help onboard contributers.
```
Comment (markkaylor) [inline docs/docs/guides/04-fe-coding-guidelines.mdx:596]: Should this content be in a more global space in the contrib docs and just linked to here? This content doesn't feel front-end specific.

**Review by markkaylor — COMMENTED** [review]
Nice work @joshuaellis, this is great.

File: docs/docs/guides/04-fe-coding-guidelines.mdx [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+don't (which is fairly typical) include it in the file you're using it. As a guide, we expect
+anything exported to be unit-tested and if a utility or hook is unique and bespoke to the component
+you're creating then it's advised to test that code in the context of said component, this is
+covered in the Testing guides below.
+
+If you do need to use the utility elsewhere then follow the same practices shown in
+[Application Structure](#application-structure) keep it as close to where it's used as possible.
+Avoid re-exporting them from a generic `index` file. However, you might find it useful to group them
+under a generic file name such as `strings.ts` which would include all the `string` utils you have.
+
```
Comment (gu-stav) [inline docs/docs/guides/04-fe-coding-guidelines.mdx:84]: Should we add here that we want to stop using re-exports from `index.ts` files and similar to lodash start removing them subsequently?

File: docs/docs/guides/04-fe-coding-guidelines.mdx [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+Whilst we are actively developing V4 of the CMS we should _always_ be using `react-query`'s hooks to
+manage the fetching and mutation of data in conjunction with `useFetchClient`. This is because
+`react-query` caches our data-fetching meaning multiple calls are not necessarily made unless a
+mutation occurs as well as handling various intermittent states and events during the fetching
+lifecycle.
+
+:::note
+We're using `react-query` version 3, see the specific docs – https://tanstack.com/query/v3/docs/react/overview
+:::
+
```
Comment (gu-stav) [inline docs/docs/guides/04-fe-coding-guidelines.mdx:220]: Could we add guidelines for the query cache key here?

- use the URL fragments from the called URL (`/admin/content-manger/:id/stages` -> `['content-manager', id, 'stages']` except `admin`
- ee calls should be prefixed with `ee`, to be able to invalidate it all at once one day
- all dynamic variables (including GET params) need to be part of the dependency array

File: docs/docs/guides/04-fe-coding-guidelines.mdx (hunk 2 above)
Comment (gu-stav) [reply docs/docs/guides/04-fe-coding-guidelines.mdx:596]: I agree.

File: docs/docs/guides/04-fe-coding-guidelines.mdx (hunk 3 above)
Comment (christiancp100) [reply docs/docs/guides/04-fe-coding-guidelines.mdx:84]: why can't we make a combination of the two things, the one you mentioned in [this comment](https://github.com/strapi/strapi/pull/17739/files#r1301163952) and re-exporting:

```shell
hooks
├─ tests
├── useContentManager.test.ts
├── useOtherHook.test.ts
├─ useContentManager.ts
├─ useOtherHook.ts
├─ index.ts
```

File: docs/docs/guides/04-fe-coding-guidelines.mdx [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+  //...
+};
+```
+
+The above function isn't expensive but more importantly we don't know what impact it has without
+measuring it. Therefore, avoid using `useCallback` and `useMemo` unless you _know_ you need it.
+
+## Dependencies
+
+Generally speaking, always opt to use native-browser APIs where possible over library code, the
```
Comment (christiancp100) [inline docs/docs/guides/04-fe-coding-guidelines.mdx:134]: Does this mean we should avoid using lodash when possible? Even though the code readability might be better?

Comment (christiancp100) [reply docs/docs/guides/04-fe-coding-guidelines.mdx:134]: Ok I kept reading and it's clear. Avoid lodash

File: docs/docs/guides/04-fe-coding-guidelines.mdx (hunk 1 above)
Comment (joshuaellis) [reply docs/docs/guides/04-fe-coding-guidelines.mdx:334]: I think you're right, lets not advertise this it's confusing ✅

File: docs/docs/guides/04-fe-coding-guidelines.mdx (hunk 3 above)
Comment (joshuaellis) [reply docs/docs/guides/04-fe-coding-guidelines.mdx:84]: I think that's clear with "Avoid re-exporting them from a generic `index` file", i don't think we need to add everywhere we're trying to remove this XYZ, it's more implied imo?

**Comment by joshuaellis** [comment]
> This is really good thank you! do you think it's worth mentioning our approach to loading in EE compos? e.g.
> 
> ```
>   const Information = useEnterprise(
>     InformationBoxCE,
>     async () =>
>       (await import('../../../../../ee/admin/content-manager/pages/EditView/InformationBox'))
>         .InformationBoxEE
>   );
> ```

I think that's more of an implementation detail? I mean the old way wouldn't work anyway since we removed `babel` 🤔 

what benefits do you see? 😄

**Review by gu-stav — APPROVED** [review]

**Review by christiancp100 — APPROVED** [review]
Good job!

**Review by jhoward1994 — APPROVED** [review]

**Pull request merged**